lib/recursive-methods.rb

```diff
@@ -3,47 +3,62 @@
 # Time complexity: ?
 # Space complexity: ?
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif (n == 1) || (n== 0)
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
 # Time complexity: ?
 # Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+def reverse(s) 
```

Comment:
👍 
This works, but because you create a new array with each recursive call this is O(n<sup>2</sup>) for both time/space complexity.